### PR TITLE
Add support for async readout of decode outputs in `tt_transformers`

### LINKED ISSUE #26341: Add support for async readout of decode outputs in `tt_transformers`
For data-parallel models `process_output_decode` takes up significant amount of time. We can batch decode iterations and overlap decode forward with reading tensors from device like it was done for llama TG here https://github.com/tenstorrent/tt-metal/pull/24707

### PR DESCRIPTION
**Pull request by rdraskicTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/26341

### Problem description
We want to support async readout to activate existing vllm feature.
https://github.com/tenstorrent/vllm/pull/162

### Checklist
- [x] [all post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17076863078)
- [x] [llama 8B DP=4 on TG](https://github.com/tenstorrent/tt-shield/actions/runs/16994089907)
- [x] [vllm nightly](https://github.com/tenstorrent/tt-metal/actions/runs/17067674360)
- [x] [galaxy demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/17065752682)

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -642,10 +643,27 @@ def decode_forward(
...
+        return host_outputs, read_events
+
+    # Note: This function is called by vLLM
+    def read_decode_output(self, tt_out, read_from_device=False, is_tokens=False):
+        """
+        Converts the input ttnn tensor to a torch tensor. The input can be logits (if is_tokens=False) or tokens (if is_tokens=True).
+        If read_from_device=False, the input is expected to be a host tensor; otherwise, it is a device tensor.
+        """
+        if read_from_device:
+            tt_out = [out.cpu() for out in tt_out]
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Linked comment in corresponding vLLM PR - https://github.com/tenstorrent/vllm/pull/162#discussion_r2283336072

File: models/demos/llama3_70b_galaxy/tt/generator.py [hunk 2]
```diff
@@ -486,9 +486,15 @@ def _easy_trace_text(
...
-        return logits, read_event
+    def read_decode_output(self, tt_out, async_read=True):
+        if not async_read:
+            return tt_out.cpu()
+
+        logits, read_event = self.model.process_output_decode(tt_out)
+        return logits, [read_event]
+
+    def process_decode_output_host(self, tt_out, is_tokens=True):
+        return ttnn.to_torch(ttnn.get_device_tensors(tt_out)[0])[0, 0, 0, :]
```
Comment (skhorasganiTT) [inline models/demos/llama3_70b_galaxy/tt/generator.py:497]: Slice to 32 on last dim was removed? Is that not needed?

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -636,15 +636,32 @@ def decode_forward(
...
+            read_events.append(ttnn.record_event(self.model[i].mesh_device, 0))
+
+        return host_outputs, read_events
+
+    # Note: This function is called by vLLM
+    def process_decode_output_host(self, tt_out, is_tokens=False):
         """
-        Input is ttnn device tensor of logits if is_tokens=False, otherwise tokens. Output is the corresponding torch tensor.
+        Converts the input ttnn host tensor to a torch tensor. The input can be logits (if is_tokens=False) or tokens (if is_tokens=True).
+        If read_from_device=False, the input is expected to be a host tensor; otherwise, it is a device tensor.
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Obsolete comment about read_from_device flag

File: models/tt_transformers/tt/model.py [hunk 4]
```diff
@@ -274,7 +274,7 @@ def concat_device_output(self, tt_out):
         """
         Concatenate the output of the devices into a single tensor.
         """
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/model.py:276]: Would be good to have a comment saying inputs are on host. Also maybe fcn name is misleading now since it has device in the name.

**Review by skhorasganiTT — APPROVED** [review]

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 2 above)
Comment (rdraskicTT) [reply models/demos/llama3_70b_galaxy/tt/generator.py:497]: It looks like it's not needed. By definition it shouldn't exceed 32 so it looked redundant. Also looked more consistent to have it like this since it was written this way in vllm.

**Review by yieldthought — APPROVED** [review]
🆗 if tests pass and Salar's comments are addressed

**Review by djordje-tt — APPROVED** [review]
